server/serverimpl.go

```diff
@@ -246,14 +263,36 @@ func (s *server) handleWSConnection(reqCtx context.Context, wsConn *websocket.Co
 	connectionCallbacks.OnConnected(reqCtx, agentConn)
 
 	sentCustomCapabilities := false
+	serverCtx := s.serverCtx
+	if serverCtx == nil {
+		// if serverCtx is nil, use background context
+		// this is Attach() case without Start() call
+		serverCtx = context.Background()
+	}
+
+	msgCh := runReadLoop(serverCtx, wsConn, &readLoopWg)
 
 	// Loop until fail to read from the WebSocket connection.
 	for {
-		msgContext := context.Background()
+		msgContext := s.serverCtx
 		request := protobufs.AgentToServer{}
 
-		// Block until the next message can be read.
-		mt, msgBytes, err := wsConn.ReadMessage()
+		var mt int
+		var msgBytes []byte
+		var err error
+
+		// Wait for the message to be read or for the server to shut down.
+		select {
+		case <-serverCtx.Done():
+			s.logger.Debugf(msgContext, "Server is shutting down.: %v", s.serverCtx.Err())
+			return
+		case msg, ok := <-msgCh:
```

Comment:
@andykellr can you take a look at this? I wonder what the performance overhead is for communicating all messages via a channel. Do you anticipate a significant overhead at large scale / large number of connection and messages?